sound/soc/sof/ipc4-pcm.c

```diff
@@ -462,7 +462,91 @@ static int sof_ipc4_pcm_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm
 	return 0;
 }
 
+static void sof_ipc4_build_time_info(struct snd_sof_dev *sdev, struct snd_sof_pcm_stream *spcm)
+{
+	struct sof_ipc4_copier *host_copier = NULL;
+	struct sof_ipc4_copier *dai_copier = NULL;
+	struct sof_ipc4_llp_reading_slot llp_slot;
+	struct sof_ipc4_timestamp_info *info;
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_dai *dai;
+	int i;
+
+	/* find host & dai to locate info in memory window */
+	for_each_dapm_widgets(spcm->list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+
+		if (!swidget)
+			continue;
+
+		if (WIDGET_IS_AIF(swidget->widget->id)) {
+			host_copier = swidget->private;
+		} else if (WIDGET_IS_DAI(swidget->widget->id)) {
+			dai = swidget->private;
+			dai_copier = dai->private;
+		}
+	}
+
+	info = spcm->private;
+	info->host_copier = host_copier;
+	info->dai_copier = dai_copier;
+	info->llp_offset = offsetof(struct sof_ipc4_fw_registers, llp_gpdma_reading_slots) +
+				    sdev->fw_info_box.offset;
+
+	/* find llp slot used by current dai */
+	for (i = 0; i < SOF_IPC4_MAX_LLP_GPDMA_READING_SLOTS; i++) {
+		sof_mailbox_read(sdev, info->llp_offset, &llp_slot, sizeof(llp_slot));
+		if (llp_slot.node_id == dai_copier->data.gtw_cfg.node_id)
+			break;
+	}
+
+	if (i == SOF_IPC4_MAX_LLP_GPDMA_READING_SLOTS) {
+		/* if no llp gpdma slot is used, check aggregated sdw slot */
+		info->llp_offset = offsetof(struct sof_ipc4_fw_registers, llp_sndw_reading_slots) +
+						sdev->fw_info_box.offset;
+		for (i = 0; i < SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS; i++) {
+			sof_mailbox_read(sdev, info->llp_offset, &llp_slot, sizeof(llp_slot));
+			if (llp_slot.node_id == dai_copier->data.gtw_cfg.node_id)
+				break;
+		}
+
+		/* no llp support */
+		if (i == SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS) {
+			dev_info(sdev->dev, "no llp found, fall back to default path");
+			info->llp_offset = 0;
+			return;
+		}
+	}
+
+	info->llp_offset += i * sizeof(struct sof_ipc4_llp_reading_slot);
+}
+
+static int sof_ipc4_pcm_hw_params(struct snd_soc_component *component,
+				  struct snd_pcm_substream *substream,
+				  struct snd_pcm_hw_params *params,
+				  struct snd_sof_platform_stream_params *platform_params)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct sof_ipc4_timestamp_info *time_info;
+	struct snd_sof_pcm *spcm;
+
+	spcm = snd_sof_find_spcm_dai(component, rtd);
+	time_info = spcm->stream[substream->stream].private;
+	/* delay calculation is not supported by current fw_reg ABI*/
+	if (!time_info)
+		return 0;
+
+	time_info->stream_start_offset = SOF_IPC4_INVALID_STREAM_POSITION;
+	time_info->llp_offset = 0;
+
+	sof_ipc4_build_time_info(sdev, &spcm->stream[substream->stream]);
+
+	return 0;
+}
+
 const struct sof_ipc_pcm_ops ipc4_pcm_ops = {
+	.hw_params = sof_ipc4_pcm_hw_params,
```

Comment:
this is an awkward place as it has nothing to do with `hw_params` callback, it is just resetting things, should this be in `prepare` or handled in `trigger`?

Reply:
Or in `pcm_setup`?

Reply:
sof_pcm_prepare will finally call sof_pcm_hw_params.  A stream can be triggered several times and but the time_info is initialized only once.  It is initialized in hw_params since all the modules are set up in FW by host driver before this function, so we can get node information in fw_reg. And modules are not set up when pcm_setup is called, so we can't get node info in fw_reg.

Reply:
@RanderWang , you are aware that hw_params can also be called several times?
In a normal playback this hw_params will be called twice, right? First during the ALSA hw_params then during the ALSA prepare.
Is it OK?

Reply:
@ujfalusi  we check the following flags for the second hw_params, So no problem
````c
	if (spcm->prepared[substream->stream])
		return 0;
````
or 
````c
	/*
	 * Handle repeated calls to hw_params() without free_pcm() in
	 * between. At least ALSA OSS emulation depends on this.
	 */
	if (pcm_ops && pcm_ops->hw_free && spcm->prepared[substream->stream]) {
		ret = pcm_ops->hw_free(component, substream);
		if (ret < 0)
			return ret;

		spcm->prepared[substream->stream] = false;
	}
````